0-common/scala-core/src/main/scala/com.snowplowanalytics.iglu/core/SelfDescribed.scala

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) 2016 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.iglu
+package core
+
+// Scala
+import scala.language.implicitConversions
+
+/**
+ * Type-class representing essence of Self-describing data
+ * It can be instantiated for any type [[E]] which is potentially
+ * containing its own description (as [[SchemaKey]])
+ *
+ * @tparam E entity type, mostly intended for various JSON ADTs,
+ *           like Json4s, Jackson, circe, Argonaut etc,
+ *           but also can be anything that can bear reference to
+ *           its description like Thrift, Map[String, String] etc
+ */
+trait SelfDescribed[E] {
```

Comment:
Yes, there's a reason. My initial plan was to add two type classes: `SelfDescribing` and `SelfDescribed`. 

Below is kind of algebra of self-description. Long story short: `SelfDescribing` useful for creating Self-describing *something*, for example in trackers, and `SelfDescribed` useful for decomposing it, for example in Scala Common Enrich.

**`SelfDescribing`** here is for entities which we can describe. For example, we can add `SchemaKey` to JValue ```{"longitude": 32.3, "latitude": 65.1}```, so it will become Self-describing JSON instance (still JValue) and primary method for this type class would be:
```scala
def describe[E: SelfDescribed](schemaKey: SchemaKey, entity: E): E // context-bound isn't necessary
```
`E` here will always be JValue (or any other type which can be self-described), before `describe` call and after it. Also `describe`cannot fail. It can be double-described non-sense, but not failed (not an `Option[E]`) .

**`SelfDescribed`** on the other hand if for entities which already can bear `SchemaKey`. Primary method is:
```scala
 def getSchemaKey[E](entity: E): Option[SchemaKey]
```
which can fail. We also have no idea if `E` will remain same `E` (type, not value) without `SchemaKey` as in example with [SelfDescribedString](https://github.com/snowplow/iglu/pull/138/files/d91d7c2ddfc66e0f44196f73514f2aa2a5558686#diff-be297bb846e2436bb242db98395f3548R34). It also means single type class isn't sufficient.

I didn't add another type class because at that time I thought it could be over-engineering, but after short sight on Scala Tracker I changed my mind and I think it's worth to add both.